Fix #5563 inconsistent behaviour of Status filter

sboddy opened the pull request:
Fixes for inconsistency as detailed in #5563
- Hiding last status hides last items in viewport,
- Disabling the status filter makes all status' visible again.

sboddy committed:
Fix #5563 inconsistent behaviour of Status filter

sboddy committed:
Merge branch 'IfcOpenShell:v0.8.0' into fix-5563

Andrej730 commented:
Looks great, thank you for looking into this.

A few notes:
- I think we can remove `self.report({"INFO"}, "No statuses selected.")` branch entirely as it's now insignificant
- I find the approach you chose for the disabling indeed the most intuitive one. But I think it might need a second part - `bim.enable_status_filters` to remember the last applied filters and apply them. Currently `bim.enable_status_filters` just obliterates previous status values and the user would need to reapply them manually each time.
- Now `bim.activate_status_filters` and `bim.disable_status_filters` reuse a lot of code. Would be nice to move it to a common method somewhere in `tool.Sequence`.

Unrelated to this PR, just a thing I've noticed testing this - `bim.activate_status_filters` operator is not exposed to UI and it's only used internally in a few places. 
Currently, when user have filters active and some filter goes out of date (e.g. a new element introduced that wasn't affected by filters or some older element changed status and perhaps should be hidden now), they would switch `is_visible` toggle to refresh it.
It probably would be more intuitive if we would expose `bim.activate_status_filters` as a plain refresh button - so refresh would be represented more explicitly in the UI and users would be able to add this operator to Quick Favorites to update whenever they feel like it. 

----

Anyways, I'm fine merging it the way it is now and fix those caveats later, let me know if you're interested in working on this or we should proceed with the currrent version of the PR for now.

sboddy commented:
1. OK, Will do.
2. I certainly had that thought in mind. I was just looking at fixing the basic inconsistencies initially. But I'll come up with something for this.
3. Yeah... I knew it was ugly dup. code. Just wasn't sure where to factor it out to, but you've pointed the way :-)

I'll pop this PR into draft and do a fix it up.

sboddy committed:
Merge branch 'IfcOpenShell:v0.8.0' into fix-5563

sboddy committed:
Improve fix for #5563 based on feedback:
- Remove INFO - No statuses selected,
- Remember and set last status on disable-reenable
- Moved duplicated code into shared tool.Sequence class method

sboddy committed:
Minor refinements for 931d8e6

sboddy commented:
I'm happy with this now. Please merge.

sboddy commented on src/bonsai/bonsai/bim/module/sequence/operator.py:
```diff
@@ -43,17 +43,19 @@ class EnableStatusFilters(bpy.types.Operator):
-                    statuses.update(element.EnumerationValues)
-                elif pset.Name == "EPset_Status":  # Our secret sauce
-                    statuses.update(element.EnumerationValues)
+                if element.PartOfPset and isinstance(element.EnumerationValues, tuple):
```
I've just realized I left this in this PR by mistake. I added it looking at https://github.com/IfcOpenShell/IfcOpenShell/issues/4067#issuecomment-3002194107 due to some peculiar data causing the status filter to error. I don't know if the problem was theoryshaw's file being broken somehow. Might as well leave it in as an added resilience bonus.

Andrej730 replied:
Well, technically leaving `EnumerationValues` as `None` is a valid IFC, so we might as well protect ourselves from this case.

![image](https://github.com/user-attachments/assets/9aed8f8b-5e49-46eb-bfc1-17846bd92bda)

Though no idea how it can occur - in theory in Bonsai if 0 elements are selected in Status enum, then `IFCPROPERTYENUMERATEDVALUE` removed completely.

Andrej730 commented:
Looks good! 🚢
